e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts

```diff
@@ -0,0 +1,916 @@
+ */
+describe("scenarios > dashboard > filters > query stages", () => {
+  beforeEach(() => {
+    restore();
+    cy.signInAsAdmin();
+    createBaseQuestions();
+  });
+
+  // Sanity checks. If the base queries tests fail then something is very wrong.
+  describe("base queries", () => {
+    const ORDERS_QUESTION_INDEX = 0;
+    const BASE_QUESTION_INDEX = 1;
+    const BASE_MODEL_INDEX = 2;
+
+    beforeEach(() => {
+      cy.then(function () {
+        createAndVisitDashboard([
+          this.ordersQuestion,
+          this.baseQuestion,
+          this.baseModel,
+        ]);
+      });
+    });
+
+    it("allows to map to all relevant columns", () => {
+      editDashboard();
+
+      cy.log("## date columns");
+      getFilter("Date").click();
+      verifyDateMappingOptions();
+
+      cy.log("## text columns");
+      getFilter("Text").click();
+      verifyTextMappingOptions();
+
+      cy.log("## number columns");
+      getFilter("Number").click();
+      verifyNumberMappingOptions();
+    });
+
+    function verifyDateMappingOptions() {
+      verifyDashcardMappingOptions(ORDERS_QUESTION_INDEX, [
+        ["Order", ORDERS_DATE_COLUMNS],
+        ["Product", PRODUCTS_DATE_COLUMNS],
+        ["User", PEOPLE_DATE_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_QUESTION_INDEX, [
+        ["Q0 Order", ORDERS_DATE_COLUMNS],
+        ["Product", PRODUCTS_DATE_COLUMNS],
+        ["User", PEOPLE_DATE_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_MODEL_INDEX, [
+        ["Base Orders Model", ORDERS_DATE_COLUMNS],
+        ["Product", PRODUCTS_DATE_COLUMNS],
+        ["User", PEOPLE_DATE_COLUMNS],
+      ]);
+    }
+
+    function verifyTextMappingOptions() {
+      verifyDashcardMappingOptions(ORDERS_QUESTION_INDEX, [
+        ["Product", PRODUCTS_TEXT_COLUMNS],
+        ["User", PEOPLE_TEXT_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_QUESTION_INDEX, [
+        ["Product", PRODUCTS_TEXT_COLUMNS],
+        ["User", PEOPLE_TEXT_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_MODEL_INDEX, [
+        ["Product", PRODUCTS_TEXT_COLUMNS],
+        ["User", PEOPLE_TEXT_COLUMNS],
+      ]);
+    }
+
+    function verifyNumberMappingOptions() {
+      verifyDashcardMappingOptions(ORDERS_QUESTION_INDEX, [
+        ["Order", ORDERS_NUMBER_COLUMNS],
+        ["Product", PRODUCTS_NUMBER_COLUMNS],
+        ["User", PEOPLE_NUMBER_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_QUESTION_INDEX, [
+        ["Q0 Order", ORDERS_NUMBER_COLUMNS],
+        ["Product", PRODUCTS_NUMBER_COLUMNS],
+        ["User", PEOPLE_NUMBER_COLUMNS],
+      ]);
+      verifyDashcardMappingOptions(BASE_MODEL_INDEX, [
+        ["Base Orders Model", ORDERS_NUMBER_COLUMNS],
+        ["Product", PRODUCTS_NUMBER_COLUMNS],
+        ["User", PEOPLE_NUMBER_COLUMNS],
+      ]);
+    }
+  });
+
+  describe("1-stage queries", () => {
+    describe("Q1 - join, custom column, no aggregations, no breakouts", () => {
```

Comment:
Every Q (Query) is tested against the matrix of 4 cases:
- card is a question / card is a model
- source table is a question / source table is a model